src/serac/numerics/equation_solver.cpp

```diff
@@ -5,20 +5,593 @@
 
 namespace serac {
 
+/// Newton solver with a 2-way line-search.  Reverts to regular Newton if max_line_search_iterations is set to 0.
+class NewtonSolver : public mfem::NewtonSolver {
+protected:
+  /// initial solution vector to do line-search off of
+  mutable mfem::Vector x0;
+  /// nonlinear solver options
+  NonlinearSolverOptions nonlinear_options;
+
+public:
+  /// constructor
+  NewtonSolver(const NonlinearSolverOptions& nonlinear_opts) : nonlinear_options(nonlinear_opts) {}
+
+#ifdef MFEM_USE_MPI
+  /// parallel constructor
+  NewtonSolver(MPI_Comm comm_, const NonlinearSolverOptions& nonlinear_opts)
+      : mfem::NewtonSolver(comm_), nonlinear_options(nonlinear_opts)
+  {
+  }
+#endif
+
+  /// Evaluate the residual, put in rOut and return its norm.
+  double evaluate_norm(const mfem::Vector& x, mfem::Vector& rOut) const
+  {
+    double normEval = std::numeric_limits<double>::max();
+    try {
+      oper->Mult(x, rOut);
+      normEval = Norm(rOut);
+    } catch (const std::exception&) {
+      normEval = std::numeric_limits<double>::max();
+    }
+    return normEval;
+  }
+
+  /// @overload
+  void Mult(const mfem::Vector&, mfem::Vector& x) const
+  {
+    MFEM_ASSERT(oper != NULL, "the Operator is not set (use SetOperator).");
+    MFEM_ASSERT(prec != NULL, "the Solver is not set (use SetSolver).");
+
+    using real_t = mfem::real_t;
+
+    real_t norm, norm_goal;
+    oper->Mult(x, r);
+
+    norm = initial_norm = Norm(r);
+    if (print_options.first_and_last && !print_options.iterations) {
+      mfem::out << "Newton iteration " << std::setw(3) << 0 << " : ||r|| = " << std::setw(13) << norm << "...\n";
+    }
+
+    norm_goal            = std::max(rel_tol * initial_norm, abs_tol);
+    prec->iterative_mode = false;
+
+    int it = 0;
+    for (; true; it++) {
+      MFEM_ASSERT(mfem::IsFinite(norm), "norm = " << norm);
+      if (print_options.iterations) {
+        mfem::out << "Newton iteration " << std::setw(3) << it << " : ||r|| = " << std::setw(13) << norm;
+        if (it > 0) {
+          mfem::out << ", ||r||/||r_0|| = " << std::setw(13) << norm / initial_norm;
+        }
+        mfem::out << '\n';
+      }
+
+      if (norm <= norm_goal && it >= nonlinear_options.min_iterations) {
+        converged = true;
+        break;
+      } else if (it >= max_iter) {
+        converged = false;
+        break;
+      }
+
+      real_t norm_nm1 = norm;
+
+      grad = &oper->GetGradient(x);
+      prec->SetOperator(*grad);
+
+      prec->Mult(r, c);  // c = [DF(x_i)]^{-1} [F(x_i)-b]
+
+      // there must be a better way to do this?
+      x0.SetSize(x.Size());
+      x0 = 0.0;
+      x0.Add(1.0, x);
+
+      real_t stepScale = 1.0;
+      add(x0, -stepScale, c, x);
+      norm = evaluate_norm(x, r);
+
+      const int               max_ls_iters = nonlinear_options.max_line_search_iterations;
+      static constexpr real_t reduction    = 0.5;
+
+      const double sufficientDecreaseParam = 0.0;  // 1e-15;
```

Comment:
Nitpick: the style is inconsistent. Camel case and underscores are mixed. Could you make a note to clean this up later? I know some of this is bound to change soon, anyway.